### PR TITLE
Refactored MFile state_dir_from_path and fixed a panic that could happen in rare circumstances

### PR DESCRIPTION
**Pull request by evanspearman**
There were two potential panics that could occur in `state_dir_from_path`:
1. If `$HOME` wasn't defined
2. If An mfile was in a directory with non-valid utf-8 characters somewhere in its path

For 1. we will now fall back to using the absolute path and prefixing with `-` like we do for anything not in the home directory.

The fix for 2 is a bit more complex. This method both takes a path and returns a path, so in theory we didn't need to round trip it through a Rust string and could keep it as an os string (allows non-utf8 characters) the whole time. It made the processing slightly more complex, but it's the cleanest way to handle that case.

## Summary by CodeRabbit

* **Improvements**
  * More reliable handling of path components and durable state directory computation, including better home-directory handling and dash-escaping for cross-platform consistency.

* **Tests**
  * Expanded unit tests covering path encoding behavior and edge cases to improve robustness.

[Review Change Stack](https://app.coderabbit.ai/change-stack/gominimal/minimal/pull/187?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Review by twitchyliquid64 — APPROVED** [review]
Looks good! As discussed online:

1. Basing the state dir path computation off `mfile_dir` instead of `repo_dir()` was likely an original sin, lets wire such that callers to `state_dir_from_path()` use the repo dir (aka the parent dir for when the minimal file is at `.minimal/minimal.toml`).
2. We can ignore the one-time breakage this represents, the one-time cache miss simply means it wont re-use cached incremental builds

File: crates/mfile/src/lib.rs [hunk 1]
```diff
@@ -2,6 +2,8 @@
 
 use std::collections::HashMap;
 use std::env::{self, home_dir};
+use std::ffi::{OsStr, OsString};
+use std::os::unix::ffi::{OsStrExt, OsStringExt};
```
[the remark this answers is not in the document]
Comment (norrietaylor) [reply crates/mfile/src/lib.rs:6]: Where we are going we don't need windows 😛

**Review by norrietaylor — APPROVED** [review]

**Comment by evanspearman** [comment]
Going to rebase this and do a quick re-test before merge
